app/services/puppet_fact_parser.rb

```diff
@@ -36,6 +36,9 @@ def operatingsystem
       end
     end
 
+    # Make sure the operating system object that we want to create is not a duplicate of an existing one.
+    os = Operatingsystem.find_by_attributes(name: os.name, major: os.major, minor: os.minor, description: os.description).first || os
```

Comment:
This is duplicating logic we already have and firing off more database queries than needed. I'd prefer to replace the calls of `find_or_initialize_by` with `find_by_attributes` and dealing with the fact it can return no instances.